### ISSUE TITLE
Feature request: on-the-fly masking for pretraining (MLM + RTD)

### ISSUE THREAD

**Issue by tc-wolf** [keras-team/keras-nlp#796]
**Is your feature request related to a problem? Please describe.**

In the [BERT data creation script](https://github.com/keras-team/keras-nlp/blob/master/examples/bert_pretraining/bert_create_pretraining_data.py) the masking is all done statically (before training, then write out to a TF Records file).  This means that on each epoch, the same tokens will be masked for any given input, so if training multiple epochs this is less data-efficient than it could be.

**Describe the solution you'd like**

Use an on-the-fly "data collator" approach like [Hugging Face](https://github.com/huggingface/transformers/blob/v4.26.1/src/transformers/data/data_collator.py#L609) or [the TF model garden (with TF Text ops)](https://github.com/tensorflow/models/blob/master/official/nlp/data/pretrain_text_dataloader.py) use so that masking is done as the dataset is read.

This would also be useful for https://github.com/keras-team/keras-nlp/issues/794 as ELECTRA pretraining has a masked language modeling objective for the 'generator'.

This also has the advantage of not requiring a large up-front preprocessing step for large datasets.

**Describe alternatives you've considered**

The user can roll-their-own for this using `tf.data.Dataset.from_generator` and whatever arbitrary transformation, but this would help avoid a lot of duplicated effort (I've implemented something similar using HF datasets + a custom collator for ELECTRA-style pretraining in a private repo).

My implementation used `gather` / `unique_with_counts` ops (and so the preprocessing function as a tf.function couldn't be used with XLA), but the implementation in `pretrain_text_dataloader.py` ^^^ with TF Text ops might not be limited in the same way.

**Comment by Samarendra109** [comment]
I can take up the issue if approved.

**Comment by mattdangerw** [comment]
I actually think we have this! At least for MLM preprocessing.

What is currently released:
 - This layer -> https://keras.io/api/keras_nlp/layers/mlm_mask_generator/
 - A guide showing it in action -> https://keras.io/guides/keras_nlp/transformer_pretraining/

What is soon to be released (probably end of month):
 - End to end MLM tasks, like the one demoed here -> https://colab.sandbox.google.com/gist/mattdangerw/fead93c20342c70e438a253f392bd425/roberta-mlm.ipynb

Is this the sort of thing you are looking for?

For RTD, I do think it's possible we will want to add some core library functionality. Perhaps even an RTD task similar to the MLM task shown in the colab. But I think we can start with the guide that @shivance is working on and go from there.

**Comment by tc-wolf** [comment]
That is excellent, that's exactly what I was looking for re: MLM pretraining!  i think a similar layer for RTD would be very useful, though it's would be a little more complicated because it needs to:

* Take already-masked positions as input
* Do replacement of tokens at masked positions using a model with MaskedLMHead
  * Rerun (or at least label as non-replaced) if MLM replaces with original token values
